zend_abstract_interface/methods/php5/methods.c

```diff
@@ -0,0 +1,113 @@
+}
+
+#ifndef NDEBUG
+static bool z_is_lower(const char *str) {
+    char *p = (char *)str;
+    while (*p) {
+        if (isalpha(*p) && !islower(*p)) return false;
+        p++;
+    }
+    return true;
+}
+#endif
+
+static bool z_call_method_without_args_ex(zval *object, zend_class_entry *ce, const char *method, size_t method_len,
+                                          zval **retval TSRMLS_DC) {
+    if (!ce || !method || !method_len || !retval) return false;
+
+    /* Prevent a potential dangling pointer in case the caller accidentally
+     * sent in an allocated retval.
+     */
+    if (*retval != NULL) return false;
+
+    /* If 'object' is NULL, the caller wants to call this method statically. */
+    zval **obj = object ? &object : NULL;
+
+    /* This one gets me all the time. Trying to save myself 5 minutes for the
+     * next time it inevitably happens.
+     */
+    assert(z_is_lower(method) && "Don't forget to send those method names in all lowercase. ;)");
+
+    /* There is an important ZEND_ACC_ALLOW_STATIC check that occurs within the
+     * VM that is circumvented when calling a method outside of a VM context
+     * via zend_call_method(). Bypassing this check can result in an
+     * application crash therefore we must look up the function handler and
+     * perform the check manually here.
+     *
+     * https://github.com/php/php-src/blob/PHP-5.4/Zend/zend_vm_def.h#L2639-L2642
+     */
+    zend_function *func = NULL;
+    if (zend_hash_find(&ce->function_table, method, (method_len + 1), (void **)&func) == FAILURE) return false;
+
+    /* Calling a non-static method statically can result in a SIGSEGV because
+     * the VM acts as a NULL check on the object before calling the function
+     * handler and we are bypassing that check with a direct call to
+     * zend_call_method(). We only want to allow this behavior on a non-static
+     * method if the function has been explicitly marked with
+     * ZEND_ACC_ALLOW_STATIC signifying that it is safe to do so.
+     */
+    if (!obj && func->common.scope && !(func->common.fn_flags & (ZEND_ACC_STATIC | ZEND_ACC_ALLOW_STATIC))) {
+        // "Non-static method %s::%s() cannot be called statically"
+        return false;
+    }
+
+    /* The ZEND_ACC_ABSTRACT check still occurs when calling zend_call_method()
+     * directly (but not until zend_call_function() is called). Since we
+     * already did the work of pulling out the function handler, we might as
+     * well fail early and avoid a potential zend_bailout while we can.
+     *
+     * https://github.com/php/php-src/blob/PHP-5.4/Zend/zend_execute_API.c#L838-L840
+     */
+    if (func->common.fn_flags & ZEND_ACC_ABSTRACT) {
+        // "Cannot call abstract method %s::%s()"
+        return false;
+    }
+
+    bool ret = false;
+
+    zai_sandbox sandbox;
+    zai_sandbox_open(&sandbox);
+
+    zend_try {
```

Comment:
Couldn't this be a problem if we catch bailouts in tracing closure callbacks? I assume bailouts happen because they can't recover, so does this need to "re-throw" the bailout? I'm thinking about `ext/soap` or something. I don't know, we don't intend to call any such functions, do we? Does `exit` use bailout on these versions?

Reply:
Hum. I think you're right that we cannot safely catch arbitrary zend_bailouts and go on pretending like nothing happened. In the `ext/soap` case, zend_bailouts are [caught and converted into exceptions](https://github.com/php/php-src/blob/master/ext/zend_test/tests/observer_error_04.phpt), but as you pointed out that's not the case for `exit`. (In PHP 8 `exit` is a clean shutdown thanks to the uncatchable Chuck Norris exception, `UnwindExit`. 😄) But we can expect an arbitrary unclean shutdown here when `exit` is used or some fatal error. If we ignored it, this would lead to a lot of ZMM leaks and likely a number of real memory leaks.

So although it is technically a "hole in the sandbox" to remove the `zend_try` here, it's actually safer to not catch them. If we wanted a true sandbox, we could run these as "sub requests" like [`ext/sandbox` does](https://github.com/krakjoe/sandbox/blob/7930851/src/sandbox.c#L353-L412). But I'm not sure the overhead of something like this would be worth it - especially since we're reducing our C-to-PHP seams.

Very nice _catch_ on this (pun intended). I'll remove this and update the tests.

Reply:
Do we need to remove it, or do we need to catch it, undo the sandbox and then re-throw?